config/coordinator.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"crypto/md5"
+	"encoding/binary"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+// Coordinator coordinates Alertmanager configurations beyond the lifetime of a
+// single configuration.
+// TODO: Make Coordinator thread safe
```

Comment:
Is that necessary? Isn't it the advantage of the coordinator with the subscription model that you don't need that anymore?

Reply:
Say there are two reloads triggered via the API, this would require write access to `config *Config` in a concurrent fassion. That should be synchronized, right?

Reply:
Yes. "Make Coordinator thread safe" made me think you want to somehow enable concurrent reloads.

I think we either can protect the `Reload` method by a mutex, or we can document that `Reload` must only be called from a single goroutine. (I could imagine the latter is naturally the case, but if not, a mutex might be the most convenient solution.) If that's what you meant anyway, I'd suggest to move this TODO to the `Reload` method, reworded as "protect this method with a mutex" and add to the doc comment (for now) that `Reload` is not concurrency-safe.

Reply:
Given that configuration reloading is not a super time sensitive thing and does not need to be parallelized I added `Mutex` to protect `subscribers` and `config`. Let me know what you think.